source/common/router/header_parser.h

```diff
@@ -14,54 +14,40 @@
 namespace Envoy {
 namespace Router {
 
-class RequestHeaderParser;
-typedef std::unique_ptr<RequestHeaderParser> RequestHeaderParserPtr;
+class HeaderParser;
+typedef std::unique_ptr<HeaderParser> HeaderParserPtr;
 
 /**
- * This class holds the parsing logic required during configuration build and
- * also perform evaluation for the variables at runtime.
+ * HeaderParser manipulates Http::HeaderMap instances. Headers to be added are pre-parsed to select
+ * between a constant value implementation and a dynamic value implementation based on
+ * AccessLog::RequestInfo fields.
  */
-class RequestHeaderParser {
+class HeaderParser {
 public:
-  virtual ~RequestHeaderParser() {}
-
-  static RequestHeaderParserPtr
-  parse(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers);
-
-  void evaluateRequestHeaders(Http::HeaderMap& headers,
-                              const AccessLog::RequestInfo& request_info) const;
-
-private:
-  std::list<std::pair<Http::LowerCaseString, HeaderFormatterPtr>> header_formatters_;
-
-  static HeaderFormatterPtr parseInternal(const std::string& format, const bool append);
-};
-
-class ResponseHeaderParser;
-typedef std::unique_ptr<ResponseHeaderParser> ResponseHeaderParserPtr;
-
-/**
- * This class holds the logic required to apply response headers additions and removals.
- */
-class ResponseHeaderParser {
-public:
-  virtual ~ResponseHeaderParser() {}
-
-  static ResponseHeaderParserPtr
-  parse(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers_to_add,
-        const Protobuf::RepeatedPtrField<ProtobufTypes::String>& headers_to_remove);
-
-  void evaluateResponseHeaders(Http::HeaderMap& headers) const;
+  /*
+   * @param headers_to_add defines the headers to add during calls to evaluateHeaders
+   * @return HeaderParserPtr a configured HeaderParserPtr
+   */
+  static HeaderParserPtr
+  configure(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers_to_add);
+
+  /*
+   * @param headers_to_add defines headers to add during calls to evaluateHeaders
+   * @param headers_to_remove defines headers to remove during calls to evaluateHeaders
+   * @return HeaderParserPtr a configured HeaderParserPtr
+   */
+  static HeaderParserPtr
+  configure(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers_to_add,
+            const Protobuf::RepeatedPtrField<ProtobufTypes::String>& headers_to_remove);
+
+  void evaluateHeaders(Http::HeaderMap& headers, const AccessLog::RequestInfo& request_info) const;
+
+protected:
+  HeaderParser() {}
 
 private:
-  struct HeaderAddition {
-    const Http::LowerCaseString header_;
-    const std::string value_;
-    const bool append_;
-  };
-
-  std::list<HeaderAddition> headers_to_add_;
-  std::list<Http::LowerCaseString> headers_to_remove_;
+  std::vector<std::pair<Http::LowerCaseString, HeaderFormatterPtr>> headers_to_add_;
```

Comment:
Nit: structs are somewhat more readable than `std::pair`, but this is probably debatable.